### ISSUE TITLE
x509: certificate signed by unknown authority when I use private certificate

### ISSUE THREAD

**Issue by darcyllingyan** [kubernetes/cloud-provider-openstack#499]
When I create the PVC, the PVC is Pending status, and after check the log, it shows as below:

Based on the log, it reports the "x509: certificate signed by unknown authority", as now the openstack certificate we used isn't signed by x509 and signed private, what can I do to resolve the issue? Thanks!

```
# # kubectl get pod -n kube-system
NAME                                           READY   STATUS    RESTARTS   AGE
csi-attacher-cinderplugin-0                    2/2     Running   0          40m
csi-nodeplugin-cinderplugin-2zq9d              2/2     Running   0          3h42m
csi-provisioner-cinderplugin-0                 2/2     Running   0          3h39m

#cat pvc.yaml 
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: csi-pvc-cinderplugin
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  storageClassName: csi-sc-cinderplugin
# kubectl apply -f pvc.yaml 
persistentvolumeclaim/csi-pvc-cinderplugin created

# kubectl get pvc
NAME                   STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS          AGE
csi-pvc-cinderplugin   Pending                                      csi-sc-cinderplugin   7s
# kubectl describe pvc csi-pvc-cinderplugin
Events:
  Type       Reason                Age               From                                                                                  Message
  ----       ------                ----              ----                                                                                  -------
  Normal     ExternalProvisioning  4s (x4 over 17s)  persistentvolume-controller                                                           waiting for a volume to be created, either by external provisioner "csi-cinderplugin" or manually created by system administrator
  Normal     Provisioning          1s (x3 over 17s)  csi-cinderplugin_csi-provisioner-cinderplugin-0_4bb1a2a3-34cc-11e9-8cc5-b27933f5fb30  External provisioner is provisioning volume for claim "default/csi-pvc-cinderplugin"
  Warning    ProvisioningFailed    1s (x3 over 16s)  csi-cinderplugin_csi-provisioner-cinderplugin-0_4bb1a2a3-34cc-11e9-8cc5-b27933f5fb30  failed to provision volume with StorageClass "csi-sc-cinderplugin": rpc error: code = Unknown desc = Post https://10.75.225.36:13000/v3/auth/tokens: x509: certificate signed by unknown authority
Mounted By:  nginx
```

**Comment by ramineni** [comment]
@darcyllingyan you could try with latest image now, as PR is merged . Remove any existing cinder-csi-plugin latest image in local , because if its already available it wont pull freshly from docker registry.

**Comment by darcyllingyan** [comment]
@ramineni ,
Thanks very much for the quick response.
After tested it, the PVC was created successfully.
```
# kubectl get pvc
NAME                     STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS          AGE
csi-pvc-cinderplugin     Bound    pvc-00f6dddd-3994-11e9-91ba-fa163e02d826   1Gi        RWO            csi-sc-cinderplugin   24m
```
But when I try to create a POD which create PV based on the PVC, following error is reported:
```
# kubectl get pod -n kube-system
NAME                                                         READY   STATUS    RESTARTS   AGE
csi-attacher-cinderplugin-0                                  2/2     Running   0          5m39s
csi-nodeplugin-cinderplugin-dqzsc                            2/2     Running   0          5m40s
csi-nodeplugin-cinderplugin-k8vgc                            2/2     Running   0          5m40s
csi-nodeplugin-cinderplugin-nd5z5                            2/2     Running   0          5m40s
csi-nodeplugin-cinderplugin-rwkn5                            2/2     Running   0          5m40s
csi-provisioner-cinderplugin-0                               2/2     Running   0          5m39s
csi-snapshotter-cinder-0                                     2/2     Running   0          5m38s

# kubectl get pvc
NAME                   STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS          AGE
csi-pvc-cinderplugin   Bound    pvc-00f6dddd-3994-11e9-91ba-fa163e02d826   1Gi        RWO            csi-sc-cinderplugin   41m

# kubectl get pv
pvc-00f6dddd-3994-11e9-91ba-fa163e02d826   1Gi        RWO            Delete           Bound       default/csi-pvc-cinderplugin   csi-sc-cinderplugin            20m

# kubectl describe pod nginx
Events:
  Type     Reason              Age    From                                 Message
  ----     ------              ----   ----                                 -------
  Normal   Scheduled           2m26s  default-scheduler                    Successfully assigned default/nginx to sdc-bcmt-01-edge-worker-01
  Warning  FailedAttachVolume  35s    attachdetach-controller              AttachVolume.Attach failed for volume "pvc-00f6dddd-3994-11e9-91ba-fa163e02d826" : node "sdc-bcmt-01-edge-worker-01" has no NodeID annotation
  Warning  FailedMount         23s    kubelet, sdc-bcmt-01-edge-worker-01  Unable to mount volumes for pod "nginx_default(99becfdd-3999-11e9-91ba-fa163e02d826)": timeout expired waiting for volumes to attach or mount for pod "default"/"nginx". list of unmounted volumes=[csi-data-cinderplugin]. list of unattached volumes=[csi-data-cinderplugin default-token-n2jlc]
```

**Comment by ramineni** [comment]
@darcyllingyan could you provide us nodeplugin logs for sdc-bcmt-01-edge-worker-01 node and also attacher logs. 
Could you give us o/p of , kubectl describe node   sdc-bcmt-01-edge-worker-01

**Comment by darcyllingyan** [comment]
@ramineni ,
The attacher logs is below:
[attacher.log](https://github.com/kubernetes/cloud-provider-openstack/files/2904400/attacher.log)

The node_drive_registrar log is:
[node-driver-registrar.log](https://github.com/kubernetes/cloud-provider-openstack/files/2904409/node-driver-registrar.log)

TheThe describe node sdc-bcmt-01-edge-worker-01 log is below:
[describe_node_edge_worker_01.log](https://github.com/kubernetes/cloud-provider-openstack/files/2904412/describe_node_edge_worker_01.log)

**Comment by ramineni** [comment]
@darcyllingyan are u able to do kubectl describe csinodeinfo ?

**Comment by darcyllingyan** [comment]
@ramineni ,
The result is null 
```
# [root@sdc-bcmt-01-control-worker-02 ~]kubectl describe csinodeinfo
# [root@sdc-bcmt-01-control-worker-02 ~]#  
[root@sdc-bcmt-01-control-worker-02 ~]# kubectl describe csinodeinfo
[root@sdc-bcmt-01-control-worker-02 ~]# kubectl describe csinodeinfo -n kube-system
[root@sdc-bcmt-01-control-worker-02 ~]# kubectl get csinodeinfo
No resources found.
[root@sdc-bcmt-01-control-worker-02 ~]# kubectl get csinodeinfo -n kube-system
No resources found.

```

**Comment by ramineni** [comment]
@darcyllingyan that should be created by node-driver-registrar https://kubernetes-csi.github.io/docs/csi-node-info-object.html ideally, could you manually create crd first  kubectl create -f https://raw.githubusercontent.com/kubernetes/csi-api/master/pkg/crd/manifests/csinodeinfo.yaml and then create csi manifests

**Comment by ramineni** [comment]
@darcyllingyan could you also give me the log of node-driver-registrar container , the above log is of cinder container of nodeplugin

**Comment by darcyllingyan** [comment]
@ramineni,
It's strange, just now after replaced the new csi cinder image, I can create the PVC successfully.
After delete the csi manifests and re-create the manifests, when I try to create the PVC again, it again reports "failed to provision volume with StorageClass "csi-sc-cinderplugin": rpc error: code = Unknown desc = Post https://10.75.225.36:13000/v3/auth/tokens: x509: certificate signed by unknown authority
"

**Comment by darcyllingyan** [comment]
@ramineni ,
Below is the node-driver-registrar container log:
[Uploading node-driver-registrar1.log…]()

**Comment by ramineni** [comment]
@darcyllingyan im unable to download the log

**Comment by darcyllingyan** [comment]
@ramineni ,
One PV can't be deleted successfully, not sure whether it has impact with the "x509: certificate signed by unknown authority". 
```
# kubectl delete pv pvc-00f6dddd-3994-11e9-91ba-fa163e02d826 --force
warning: --force is ignored because --grace-period is not 0.
persistentvolume "pvc-00f6dddd-3994-11e9-91ba-fa163e02d826" deleted
```

**Comment by ramineni** [comment]
@darcyllingyan Thats strange , for safe side , you could change `imagePullPolicy: Always` in all manifests files and check

**Comment by darcyllingyan** [comment]
@ramineni ,
Thanks very much for the suggestion, now the PVC can be created successfully after change ```imagePullPolicy: Always```, but there is still the issue ```AttachVolume.Attach failed for volume "pvc-3d1d6dfb-39c3-11e9-9185-fa163e02d826" : node "sdc-bcmt-01-edge-worker-01" has no NodeID annotation```.
note: I have created the CRD csinodeinfos.csi.storage.k8s.io manually.

```
# kubectl get crd|grep csi
csidrivers.csi.storage.k8s.io                    2019-02-19T09:16:55Z
csinodeinfos.csi.storage.k8s.io                  2019-02-26T12:40:57Z

```
The related logs are below:
1. nodeplugin logs for sdc-bcmt-01-edge-worker-01 node(node-driver-registrar container)
[node-driver-registrar00.log](https://github.com/kubernetes/cloud-provider-openstack/files/2905476/node-driver-registrar00.log)
:

2. attacher logs:
[attacher00.log](https://github.com/kubernetes/cloud-provider-openstack/files/2905481/attacher00.log)

3. The output of ```kubectl describe node sdc-bcmt-01-edge-worker-01```
[describe_node.log](https://github.com/kubernetes/cloud-provider-openstack/files/2905483/describe_node.log)

Do you have any suggestions? Thanks.

**Comment by ramineni** [comment]
@darcyllingyan looks like problem is NodeId annotation is missing , that should be added by node-driver-registrar. Could you give me the o/p of kubectl describe csinodeinfo and also attach node-plugin cinder container logs , attacher cinder container logs also.

**Comment by darcyllingyan** [comment]
@ramineni ,
Thanks for the information.
Below are the logs:
node-driver-registrar cinder container log:
[nodeplugin_cinder_container.log](https://github.com/kubernetes/cloud-provider-openstack/files/2906182/nodeplugin_cinder_container.log)

The attach cinder container log:
[attacher_cinder_container.log](https://github.com/kubernetes/cloud-provider-openstack/files/2906183/attacher_cinder_container.log)

The ```kubectl describe csinodeinfo``` output is still null.
I modified the feature-gates in the /etc/kubernetes/manifests/kube-apiserver.yml and kubernetes/kubelet-config.yml:
But the ```kubectl describe csinodeinfo``` still shows null.

Thanks

**Comment by ramineni** [comment]
@darcyllingyan what is the kubernetes version you are using ?

**Comment by darcyllingyan** [comment]
@ramineni

The kubernetes version is v1.12.3:
```
# kubelet --version
Kubernetes v1.12.3
``` 
In my cluster, after modifed the```/etc/kubernetes/manifests/kube-apiserver.yml``` and ```/etc/kubernetes/kubelet-config.yml```, the feature will take effect auto and don't need to aply.

Now it still shows null for ```kubectl describe csinodeinfo```

**Comment by ramineni** [comment]
@darcyllingyan minimum requirement of node-driver-registrar is v1.13 https://github.com/kubernetes-csi/node-driver-registrar#compatibility ,

**Comment by darcyllingyan** [comment]
@ramineni ,
Thanks for the response!
I will upgrade to v1.13 and try again.

Thanks
Darcy

**Comment by darcyllingyan** [comment]
@ramineni ,
After I upgraded the kubernetes version to v1.13, now all the cinder-csi plugin pod can't be started.
The detailed description is in #514 .
DO you have any suggestions?
Thanks.

**Comment by darcyllingyan** [comment]
The csinodeinfo issue has been resolved after upgrade k83 to v1.13.
